zcash_client_sqlite/src/wallet.rs

```diff
@@ -424,32 +604,39 @@ pub(crate) fn get_transparent_receivers<P: consensus::Parameters>(
 pub(crate) fn get_legacy_transparent_address<P: consensus::Parameters>(
     params: &P,
     conn: &rusqlite::Connection,
-    account: AccountId,
+    account_id: AccountId,
 ) -> Result<Option<(TransparentAddress, NonHardenedChildIndex)>, SqliteClientError> {
-    // Get the UFVK for the account.
-    let ufvk_str: Option<String> = conn
+    use zcash_address::unified::Container;
+    use zcash_primitives::legacy::keys::AccountPubKey;
+
+    // Get the UIVK for the account.
+    let uivk_str: Option<String> = conn
         .query_row(
-            "SELECT ufvk FROM accounts WHERE account = :account",
-            [u32::from(account)],
+            "SELECT uivk FROM accounts WHERE id = :account",
+            [account_id.0],
             |row| row.get(0),
         )
         .optional()?;
 
-    if let Some(ufvk_str) = ufvk_str {
-        let ufvk = UnifiedFullViewingKey::decode(params, &ufvk_str)
-            .map_err(SqliteClientError::CorruptedData)?;
+    if let Some(uivk_str) = uivk_str {
+        let (network, uivk) = Uivk::decode(&uivk_str)
+            .map_err(|e| SqliteClientError::CorruptedData(format!("Unable to parse UIVK: {e}")))?;
+        if params.network_type() != network {
+            return Err(SqliteClientError::CorruptedData(
+                "Network type mismatch".to_owned(),
+            ));
+        }
 
         // Derive the default transparent address (if it wasn't already part of a derived UA).
-        ufvk.transparent()
-            .map(|tfvk| {
-                tfvk.derive_external_ivk()
-                    .map(|tivk| tivk.default_address())
-                    .map_err(SqliteClientError::HdwalletError)
-            })
-            .transpose()
-    } else {
-        Ok(None)
+        for item in uivk.items() {
+            if let Ivk::P2pkh(ivk) = item {
+                let pubkey = AccountPubKey::deserialize(&ivk)?;
+                return Ok(Some(pubkey.derive_external_ivk()?.default_address()));
```

Comment:
Bug: the UIVK contains the transparent external IVK. This is related to the earlier bug; for UFVKs this line "cancels out" the earlier bug, but for imported UIVKs this will derive the wrong address. The old code that this is replacing used a UFVK here, so this derivation path was correct (only) in that context.